precert/corr.go

```diff
@@ -0,0 +1,202 @@
+package precert
+
+import (
+	"bytes"
+	encoding_asn1 "encoding/asn1"
+	"errors"
+	"fmt"
+
+	"golang.org/x/crypto/cryptobyte"
+	"golang.org/x/crypto/cryptobyte/asn1"
+)
+
+// Correspond returns nil if the two certificates are a valid precertificate/final certificate pair.
+// Order of the arguments matters: the precertificate is first and the final certificate is second.
+// Note that RFC 6962 allows the precertificate and final certificate to have different Issuers, but
+// this function rejects such pairs.
+func Correspond(precertDER, finalDER []byte) error {
+	preTBS, err := tbsDERFromCertDER(precertDER)
+	if err != nil {
+		return fmt.Errorf("parsing precert: %w", err)
+	}
+
+	finalTBS, err := tbsDERFromCertDER(finalDER)
+	if err != nil {
+		return fmt.Errorf("parsing final cert: %w", err)
+	}
+
+	// The first 7 fields of TBSCertificate must be byte-for-byte identical.
+	// The next 2 fields (issuerUniqueID and subjectUniqueID) are forbidden
+	// by the Baseline Requirements so we assume they are not present (if they
+	// are, they will fail the next check, for extensions).
```

Comment:
Totally not necessary, but to me it feels like it would be slightly nicer to explicitly check the field tag on the next field when we get to this point. If it's [1] or [2], bail out with an error. If it's [3] like we expect, proceed. That way the error can be a bit more specific. The [PeekASN1Tag](https://pkg.go.dev/golang.org/x/crypto/cryptobyte#String.PeekASN1Tag) method makes this kind of checking pretty easy.

Reply:
We do actually check the field tag in unwrapExtensions:

    	if !field.ReadASN1(&extensions, asn1.Tag(3).Constructed().ContextSpecific()) {

I'd rather avoid checking it in two places because then we can never exercise both conditionals in a test. Though it occurs to me I should actually write a test that exercises this conditional...

Reply:
Here's a test that exercises the first conditional of `unwrapExtensions`. The validPrecertBytes come from a processed `testdata/good/precert.pem`.
```
func TestUnwrapExtensions(t *testing.T) {
	var validPrecertBytes cryptobyte.String
	validPrecertBytes = []byte{163, 130, 1, 38, 48, 130, 1, 34, 48, 14, 6, 3, 85, 29, 15, 1, 1, 255, 4, 4, 3, 2, 5, 160, 48, 29, 6, 3, 85, 29, 37, 4, 22, 48, 20, 6, 8, 43, 6, 1, 5, 5, 7, 3, 1, 6, 8, 43, 6, 1, 5, 5, 7, 3, 2, 48, 12, 6, 3, 85, 29, 19, 1, 1, 255, 4, 2, 48, 0, 48, 29, 6, 3, 85, 29, 14, 4, 22, 4, 20, 38, 19, 204, 205, 94, 7, 254, 116, 233, 144, 222, 65, 91, 222, 158, 49, 127, 57, 203, 141, 48, 31, 6, 3, 85, 29, 35, 4, 24, 48, 22, 128, 20, 20, 46, 179, 23, 183, 88, 86, 203, 174, 80, 9, 64, 230, 31, 175, 157, 139, 20, 194, 198, 48, 85, 6, 8, 43, 6, 1, 5, 5, 7, 1, 1, 4, 73, 48, 71, 48, 33, 6, 8, 43, 6, 1, 5, 5, 7, 48, 1, 134, 21, 104, 116, 116, 112, 58, 47, 47, 114, 51, 46, 111, 46, 108, 101, 110, 99, 114, 46, 111, 114, 103, 48, 34, 6, 8, 43, 6, 1, 5, 5, 7, 48, 2, 134, 22, 104, 116, 116, 112, 58, 47, 47, 114, 51, 46, 105, 46, 108, 101, 110, 99, 114, 46, 111, 114, 103, 47, 48, 34, 6, 3, 85, 29, 17, 4, 27, 48, 25, 130, 23, 104, 111, 117, 115, 101, 116, 114, 97, 105, 110, 105, 110, 103, 112, 117, 112, 112, 121, 46, 105, 110, 102, 111, 48, 19, 6, 3, 85, 29, 32, 4, 12, 48, 10, 48, 8, 6, 6, 103, 129, 12, 1, 2, 1, 48, 19, 6, 10, 43, 6, 1, 4, 1, 214, 121, 2, 4, 3, 1, 1, 255, 4, 2, 5, 0}
	_, err := unwrapExtensions(validPrecertBytes)
	if err != nil {
		t.Error("should not have errored")
	}

	var invalidPrecertBytes cryptobyte.String
	invalidPrecertBytes = []byte{6, 6, 6}
	_, err = unwrapExtensions(invalidPrecertBytes)
	if err == nil {
		t.Error("should have errored")
	}
}
```

Reply:
Thanks for noticing I forgot to add the test, Phil! And even providing an example test case. 💪🏻 I pushed that in the latest commit.

BTW feel free to unresolve a conversation if you think I resolved it in error.